lib/customisation.nix

```diff
@@ -131,7 +131,10 @@ rec {
       origArgs = auto // args;
       pkgs = f origArgs;
       mkAttrOverridable = name: _: makeOverridable (newArgs: (f newArgs).${name}) origArgs;
-    in lib.mapAttrs mkAttrOverridable pkgs;
+    in
+      if lib.isDerivation pkgs then throw
+        "function callPackages called on a *single* derivation `${pkgs.name or "<unknown-name>"}`."
+      else lib.mapAttrs mkAttrOverridable pkgs;
```

Comment:
Parenthetical to this PR but doing a `mapAttrs` over all of `pkgs` seems like a really bad idea for performance. Hadn't noticed that before.

Reply:
Typically `callPackages` is used on tiny sets, fortunately.

Is there a faster way?  Even if it's very hacky, we can at least use it to measure potential performance benefit &ndash; estimate how much human work is worth putting into this.

Reply:
A quick `nix-env -qa ...` trace showed only a few larger sets, all under 100
```
trace: 28
trace: 34
trace: 43
trace: 43
trace: 88
trace: 93
```
(caveat is that this quick test recurses into a bit fewer sets than Hydra)

Reply:
I'll assume it's not bad enough to invest more time (of mine).